railgun/manager/manager.go

```diff
@@ -94,113 +170,119 @@ func Run(ctx context.Context, c *Config) error {
+			},
+		},
+		{
+			IsEnabled: &c.KongClusterPluginEnabled,
+			Controller: &kongctrl.KongClusterPluginReconciler{
+				Client: mgr.GetClient(),
+				Log:    ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
+				Scheme: mgr.GetScheme(),
+			},
+		},
+		{
+			IsEnabled: &c.KongPluginEnabled,
+			Controller: &kongctrl.KongPluginReconciler{
+				Client: mgr.GetClient(),
+				Log:    ctrl.Log.WithName("controllers").WithName("KongPlugin"),
+				Scheme: mgr.GetScheme(),
+			},
+		},
+		{
+			IsEnabled: &c.KongConsumerEnabled,
+			Controller: &kongctrl.KongConsumerReconciler{
+				Client: mgr.GetClient(),
+				Log:    ctrl.Log.WithName("controllers").WithName("KongConsumer"),
+				Scheme: mgr.GetScheme(),
 			},
 		},
-	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "Secret")
-		return err
-	}
-
-	// TODO - we've got a couple places in here and below where we "short circuit" controllers if the relevant API isn't available.
-	// This is convenient for testing, but maintainers should reconsider this before we release KIC 2.0.
-	// SEE: https://github.com/Kong/kubernetes-ingress-controller/issues/1101
-	if err := kongctrl.SetupIngressControllers(mgr); err != nil {
-		setupLog.Error(err, "unable to create controllers", "controllers", "Ingress")
-		return err
 	}
 
-	// TODO - similar to above, we're short circuiting here. It's convenient, but let's discuss if this is what we want ultimately.
-	// SEE: https://github.com/Kong/kubernetes-ingress-controller/issues/1101
-	udpIngressAvailable, err := kongctrl.IsAPIAvailable(mgr, &v1alpha1.UDPIngress{})
-	if !udpIngressAvailable {
-		setupLog.Error(err, "API configuration.konghq.com/v1alpha1/UDPIngress is not available, skipping controller")
-	} else {
-		if err = (&kongctrl.KongV1Alpha1UDPIngressReconciler{
-			Client: mgr.GetClient(),
-			Log:    ctrl.Log.WithName("controllers").WithName("UDPIngress"),
-			Scheme: mgr.GetScheme(),
-		}).SetupWithManager(mgr); err != nil {
-			setupLog.Error(err, "unable to create controller", "controller", "UDPIngress")
-			return err
+	for _, c := range controllers {
+		if err := c.MaybeSetupWithManager(mgr); err != nil {
+			return fmt.Errorf("unable to create controller %q: %w", c.Name(), err)
 		}
 	}
 
-	tcpIngressAvailable, err := kongctrl.IsAPIAvailable(mgr, &configurationv1beta1.TCPIngress{})
-	if !tcpIngressAvailable {
-		setupLog.Error(err, "API configuration.konghq.com/v1alpha1/TCPIngress is not available, skipping controller")
-	} else {
-		if err = (&kongctrl.KongV1Beta1TCPIngressReconciler{
-			Client: mgr.GetClient(),
-			Log:    ctrl.Log.WithName("controllers").WithName("TCPIngress"),
-			Scheme: mgr.GetScheme(),
-		}).SetupWithManager(mgr); err != nil {
-			setupLog.Error(err, "unable to create controller", "controller", "TCPIngress")
-			return err
-		}
-	}
 	// BUG: kubebuilder (at the time of writing - 3.0.0-rc.1) does not allow this tag anywhere else than main.go
```

Comment:
I'm sorry, I know this line wasn't modified in this PR but I just realized we missed adding links to the relevant upstream issues, if you feel so inclined to add those now I would like to, otherwise no worries.